### DISCUSSION TITLE
16 bit textures

### DISCUSSION THREAD

**Discussion by omri-polycam** [google/filament#7289] [Q&A]
Hey guys,
  I'm trying to use 16 bit per component textures and cannot seem to get it to work. I'm using this to encode XYZ coordinates as a `std::uint16_t` unsigned integer in each color component of each texel of an RGB texture and this texture is sampled in a custom vertex shader. The failure mode is that it seems that whenever using the `texture(...).rgb` function in the vertex shader it always evaluates to (0, 0, 0). 
   I've already verified that my shader works when encoding the coordinates as `std::uint8_t` color components in a common 8 bit per component RGB texture (`Texture::Format::RGB`, `Texture::Type::UBYTE`, `Texture::InternalFormat::RGB8`). I've also verified using a debugger that I'm passing in reasonable `std::uint16_t` encoded coordinates into the texture data.
   I've tried (`Texture::Format::RGB_INTEGER`, `Texture::Type::USHORT`, `Texture::InternalFormat::RGB16UI`) with no success. I've also tried (`Texture::Format::RGB`, `Texture::Type::HALF`, `Texture::InternalFormat::RGB16F`) with the same texture sampling always evaluating to a 0 vector no matter what data I pass into the texture pixel buffer descriptor.
   I'm currently running on MacOS with the metal backend but will also need this to work on iOS/Metal, Android/GL and WebAssembly/GL. Does anybody know if I'm passing the wrong texture format, type and internal format enumerations? Is there something special I need to do in my vertex shader? Is there something I need to set for `format` or `precision` in my `sampler2d` parameter in my material? Ideally `texture(...).rgb` evaluates to a 3D float vector with components from 0 to 1, but it's easy enough to work around if it evaluates from 0 to 65,535 as an integer. Any help is appreciated!
   
Thanks,
   Omri

**Comment by romainguy** [comment]
Filament itself uses an RGB16F texture sampled for every pixel of every lit material:

```
    Texture* lut = Texture::Builder()
            .width(DFG_LUT_SIZE)
            .height(DFG_LUT_SIZE)
            .format(backend::TextureFormat::RGB16F)
            .build(engine);

    Texture::PixelBufferDescriptor buffer(DFG_LUT, byteCount,
            Texture::Format::RGB, Texture::Type::HALF);

    lut->setImage(engine, 0, std::move(buffer));
```

This works with every backend on every platform. The sampler for it uses `Type::SAMPLER_2D, Format::FLOAT,  Precision::MEDIUM`.

**Reply by omri-polycam** [reply]
Thanks, are 16 bit bit integer components also available on every platform? This would be preferable as half precision floating point has a signifcand only marginally better than plain 8 bit and would not bias higher precision nearer to the origin/zero unlike floating point. If so, what are the type, format, internal format and precision to use it?

**Reply by romainguy** [reply]
It should work yes.

**Reply by omri-polycam** [reply]
Thanks, I got it working. One last question, for another application in the same shader, is `TextureFormat::RGB32F` available on all platforms or do I have to go with half precision?
